Add contributor ladder and roles (#1)
cilium/community: CONTRIBUTOR-LADDER.md

## Contributor Ladder

Hello! We are excited that you want to learn more about our project contributor ladder! This contributor ladder outlines the different contributor roles within the Cilium project, along with the responsibilities and privileges that come with them. Community members generally start at the first levels of the "ladder" and advance up it as their involvement in the project grows.  Our contributors are happy to help you advance along the contributor ladder.

Each of the contributor roles below is organized into lists of three types of things. "Requirements" are things that a contributor is expected to do at that level. "Qualifications" are qualifications a person needs to meet to be at that level (may be one or any/all of them), and "Privileges" are things contributors on that level are entitled to.

As the Cilium project grows, the current roles may be broken out into new roles and/or teams and roles may no longer be needed. 

The final judgment on whether an individual fulfills the criteria for a role is up to the Cilium Committers. After six months of inactivity, any contributor can be removed from their position(s).

### Community Contributor

> 

Author: ```suggestion
Description: A Community Contributor contributes directly to the project and adds value to it. Contributions need not be code. People at the Community Contributor level may be new contributors, or they may only contribute occasionally.

```

* Responsibilities:
    * Must follow the [Cilium CoC](https://github.com/cilium/cilium/blob/master/CODE_OF_CONDUCT.md)

>     * Follow the [contributing guide](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/) 

Author: ```suggestion
    * Follow the contributing guide for each project, for example see [Cilium's here](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/) 
```

>     * Follow the [contributing guide](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/) 

Author: I've made it general with Cilium as an example

> * Qualifications:
> Anybody that is participating in the Cilium community is welcome. Any contribution counts. Some possible forms of contribution are:

Reviewer: ```suggestion
* Qualifications:
  Anybody that is participating in the Cilium community is welcome. Any contribution counts. Some possible forms of contribution are:
```

An Organization Member has all the rights and responsibilities of a Community Contributor, plus:

* Responsibilities:
    * Continues to contribute regularly
    * Work to accomplish the tasks they volunteer to do within the project

* Qualifications:
    * Must have successful contributions to the project, including at least one of the following:
        * Written and/or reviewed PRs,
        * Reviewed PRs,
        * Made contributions that resolved Issues,
        * Or some equivalent combination or contribution
    * Must be actively contributing to at least one project area, where the privileges of Organization Member will be beneficial

* Privileges:
    * Can trigger CI
    * Can leave reviews on PRs
    * Can recommend other contributors to become Cilium Github Org Members
    * Can nominate themselves to other roles within Cilium

The process for a Contributor to become an Organization Member is as follows:

1. Open a PR in the member management file

> 2. Receive two “LGTM” comments from any Org Members, and approval by a committer who merges the PR

Reviewer: Nit: I'd avoid missing straight and curly quotes in the doc. Well, I'd avoid curly quotes altogether.

> 2. Receive two “LGTM” comments from any Org Members, and approval by a committer who merges the PR

Author: ```suggestion
2. Receive two "LGTM" comments from any Org Members, and approval by a committer who merges the PR
```

### Reviewer

> Description: A Reviewer has responsibility for specific code, documentation, test, or other project areas. They are collectively responsible, with other Reviewers, for reviewing all changes to those areas and indicating whether those changes are ready to merge. They have a track record of contribution and review in the project.
> Reviewers are responsible for a "specific area." This can be a specific code directory, chapter of the docs, or other clearly-defined project component that is smaller than an entire repository or subproject. Most often it is one or a set of directories in one or more Git repositories. The "specific area" below refers to this area of responsibility.
> Reviewers have all the rights and responsibilities of an Organization Member, plus:
> * Responsibilities:

Author: ```suggestion
Description: A Reviewer has responsibility for specific code, documentation, test, or other project areas. They are collectively responsible, with other Reviewers, for reviewing all changes to those areas and indicating whether those changes are ready to merge. They have a track record of contribution and review in the project.

Reviewers are responsible for a "specific area." This can be a specific code directory, chapter of the docs, or other clearly-defined project component that is smaller than an entire repository or subproject. Most often it is one or a set of directories in one or more Git repositories. The "specific area" below refers to this area of responsibility.
Reviewers have all the rights and responsibilities of an Organization Member, plus:

* Responsibilities:
```

    * Following the [reviewing guide](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#pull-requests-review-process-for-committers) 
    * Regularly reviewing Pull Requests against their specific areas of responsibility, including Pull Requests that are assigned to them
    * Helping other contributors become reviewers
    * Review CFPs and provide feedback

* Requirements:
    * Is an Organization Member
    * Has a track record of constructive and valuable reviews
    * Has demonstrated an in-depth knowledge of the specific area, for example by resolving test failures or through contributions
    * Commits to being responsible for that specific area
    * Is supportive of new and occasional contributors and helps get useful PRs in shape to commit
* Privileges:
    * Has GitHub rights to approve pull requests in specific directories
    * Can recommend and review other Org Members to become Reviewers

The process for an Organization Member to become a Reviewer is as follows:

> 1. Open a PR in the member management file for the correct team

Reviewer 2: Perhaps we should have two categories here, one that makes it easy to apply permissions for all reviewers and one for a particular code area that the org member intends to participate in?

```suggestion
1. Open a PR in the member management file for at least two teams: "Reviewers" and for a specific review area
```

2. At least two members who are already Committers, approve the PR

Automated tooling assigns PRs across the Reviewers defined for each particular area. 

### Committer

> Description: Committers have write privileges and voting privileges, and are collectively responsible for steering the project in a positive direction. As such it represents a significant level of trust in an individual’s commitment to working with other committers and the community at large for the benefit of the project. 

Reviewer: `individual’s` -> I'd also use a straight apostrophe.

Committers have all the rights and responsibilities of a Reviewer, plus:

* Responsibilities:
    * Mentoring new Committers
    * Discussing strategy and policy for the project
    * [Voting](https://docs.cilium.io/en/latest/community/governance/commit_access/#voting) on project matters when required
    * Additional responsibilities specific to any project Roles they may have
    * Approve contributors to assume roles

* Qualifications:
    * Laid out in [granting commit access](https://docs.cilium.io/en/latest/community/governance/commit_access/#granting-commit-access)
    * Is supportive of new and occasional contributors and helps get useful PRs in shape to commit
    * Is able to exercise judgment for the good of the project, independent of their employer, friends, or team
    * Mentors other contributors

> * Privileges:

Reviewer 2: ```suggestion

* Privileges:
```

    * Can merge code into the repo(s) they are contributing to
    * Represent the project in public as a Committer
    * Can communicate with the CNCF on behalf of the project, and participate in CNCF maintainer activities including votes
    * Can nominate new Committers
    * Is listed as a Committer in the MAINTAINERS.md

> Committers who are contributing and maintaining project code are automatically part of the Code Team defined in Roles.

Reviewer 2: I presume that "Roles" here is a reference to the second document?

```suggestion
Committers who are contributing and maintaining project code are automatically part of the Code Team defined in CONTRIBUTOR-ROLES.md.
```

# Contributor Ladder

* [Contributor Ladder](#contributor-ladder)
    * [Community Contributor](#community-contributor)
    * [Organization Member](#organization-member)
    * [Reviewer](#reviewer)
    * [Committer](#committer)

## Contributor Ladder

> Hello! We are excited that you want to learn more about our project contributor ladder! This contributor ladder outlines the different contributor roles within the Cilium project, along with the responsibilities and privileges that come with them. Community members generally start at the first levels of the "ladder" and advance up it as their involvement in the project grows.  Our contributors are happy to help you advance along the contributor ladder.

Reviewer 3: Nit
```suggestion
Hello! We are excited that you want to learn more about our project contributor ladder! This contributor ladder outlines the different contributor roles within the Cilium project, along with the responsibilities and privileges that come with them. Community members generally start at the first levels of the "ladder" and advance up it as their involvement in the project grows. Our contributors are happy to help you advance along the contributor ladder.
```

Author: Thanks!

Each of the contributor roles below is organized into lists of three types of things. "Requirements" are things that a contributor is expected to do at that level. "Qualifications" are qualifications a person needs to meet to be at that level (may be one or any/all of them), and "Privileges" are things contributors on that level are entitled to.

As the Cilium project grows, the current roles may be broken out into new roles and/or teams and roles may no longer be needed. 

> The final judgment on whether an individual fulfills the criteria for a role is up to the Cilium Committers. After six months of inactivity, any contributor can be removed from their position(s).

Reviewer 2: I wonder if it's also worth linking the revocation policy and clarifying that it applies to all roles as well?

https://docs.cilium.io/en/v1.13/community/governance/commit_access/#revoking-commit-access

The above is primarily written around committer access, but this could apply equally to other roles, for instance if a contributor can run CI and begins to abuse it, or if they are misusing access to community social accounts.

Author: good point, I've added "or behavior detrimental to the future of the project"

> The final judgment on whether an individual fulfills the criteria for a role is up to the Cilium Committers. After six months of inactivity, any contributor can be removed from their position(s).

Author: ```suggestion
The final judgment on whether an individual fulfills the criteria for a role is up to the Cilium Committers. After six months of inactivity or behavior detrimental to the future of the project, any contributor can be removed from their position(s).
```
